src/async_fuse/memfs/cache/storage_manager.rs

```diff
@@ -0,0 +1,812 @@
+#[derive(Debug)]
+pub struct StorageManager<S> {
+    /// The top-level storage, `InMemoryCache` for example
+    storage: Arc<S>,
+    /// Block size in bytes
+    block_size: usize,
+    /// Last modified times of the cache (on file level)
+    mtimes: HashMap<INum, SystemTime>,
+}
+
+impl<S> StorageManager<S>
+where
+    S: Storage + Send + Sync + 'static,
+{
+    /// Create a `StorageManager` with the top-level `Storage`.
+    pub fn new(storage: S, block_size: usize) -> Self {
+        StorageManager {
+            storage: Arc::new(storage),
+            block_size,
+            mtimes: HashMap::new(),
+        }
+    }
+
+    /// Convert offset in byte to block id via the equation:
+    ///
+    /// `block_id = offset / block_size`
+    fn offset_to_block_id(&self, offset: usize) -> usize {
+        offset.overflow_div(self.block_size)
+    }
+
+    /// Load blocks from the storage concurrently.
+    async fn load_blocks(&self, ino: INum, start_block: usize, end_block: usize) -> Vec<Block> {
+        let mut handles = vec![];
+
+        for block_id in start_block..end_block {
+            let storage = Arc::clone(&self.storage);
+            let handle = task::spawn(async move { storage.load(ino, block_id).await });
+            handles.push(handle);
+        }
+
+        let mut blocks = vec![];
+
+        for (handle, block_id) in handles.into_iter().zip(start_block..end_block) {
+            let block = handle
+                .await
+                .unwrap_or_else(|e| panic!("Fails when awaiting on loading: {e}"));
+            if let Some(block) = block {
+                blocks.push(block);
+            } else {
+                // A "gap" exists in the range of the being-loaded blocks,
+                // the file is considered to be truncated, and all the rest (if exist) are ignored.
+                warn!("Cannot fetch block {block_id} from storage, consider to be truncated.");
+                break;
+            }
+        }
+
+        blocks
+    }
+
+    /// Store blocks into the storage concurrently.
+    async fn store_blocks(
+        &self,
+        ino: INum,
+        start_block: usize,
+        io_blocks: impl Iterator<Item = IoBlock>,
+    ) {
+        let mut handles = vec![];
+
+        for (mut io_block, block_id) in io_blocks.zip(start_block..) {
+            io_block.set_dirty();
+            let storage = Arc::clone(&self.storage);
+            let handle = task::spawn(async move { storage.store(ino, block_id, io_block).await });
+            handles.push(handle);
+        }
+
+        for handle in handles {
+            handle
+                .await
+                .unwrap_or_else(|e| panic!("Fails when awaiting on storing: {e}"));
+        }
+    }
+
+    /// Convert `IoBlock`s from slice.
+    fn make_io_blocks_from_slice(
+        &self,
+        offset: usize,
+        data: &[u8],
+    ) -> impl Iterator<Item = IoBlock> + '_ {
```

Comment:
The implementation of this function seems a bit messy. I think we can refactor it. I feel like we don't need the abstraction of `IoBlock`; we can directly use Block and add other members inside it, along with corresponding initialization functions. If it's really necessary to use a data structure to represent the relationship between multiple blocks, I think `Slice` would be a better choice.